File: lru-cache-master/ruby/lru_cache.rb
class Node
  attr_accessor :data, :key, :next_node, :prev_node

  def initialize(data = nil, key = nil, next_node = nil, prev_node = nil)
    @data = data
    @key = key
    @next_node = next_node
    @prev_node = prev_node
  end
end

class DoublyLinkedList
  attr_reader :head, :tail

  def initialize(head = nil, tail = nil)
    @head = head
    @tail = tail
  end

  # ADD THE NODE TO THE HEAD OF THE LIST
  def add_head(node) 
    # Add the given node to the head of the list
    if @head.nil?
      @head = node
      @tail = node
    else
      node.next_node = @head
      @head.prev_node = node
      @head = node
    end
  end

  # REMOVE THE TAIL NODE FROM THE LIST
  # AND RETURN IT
  def remove_tail
    # Remove the tail node from the list and return it
    if @tail.nil?
      return nil
    else
      tail_node = @tail
      if @head == @tail
        @head = nil
        @tail = nil
      else
        @tail.prev_node.next_node = nil
        @tail = @tail.prev_node
      end
      return tail_node
    end
  end

  # REMOVE THE GIVEN NODE FROM THE LIST
  # AND THEN RETURN IT
  def remove_node(node) 
    # Remove the given node from the list and return it
    return if node.nil?

    if node == @head
      if @head == @tail
        @head = nil
        @tail = nil
      else
        @head = @head.next_node
        @head.prev_node = nil
      end
    elsif node == @tail
      @tail = @tail.prev_node
      @tail.next_node = nil
    else
      node.prev_node.next_node = node.next_node
      node.next_node.prev_node = node.prev_node
    end

    node.next_node = nil
    node.prev_node = nil
  end

  # MOVE THE GIVEN NODE FROM ITS LOCATION TO THE HEAD
  # OF THE LIST
  def move_node_to_head(node) 
    # Move the given node from its location to the head of the list
    return if node == @head

    remove_node(node)
    add_head(node)
  end
end

class LRUCache
  attr_reader :limit, :size

  def initialize(limit = 10)
    @limit = limit
    @size = 0
    @hash = {}
    @list = DoublyLinkedList.new
  end

  # RETRIEVE THE NODE FROM THE CACHE USING THE KEY
  # IF THE NODE IS IN THE CACHE, MOVE IT TO THE HEAD OF THE LIST AND RETURN IT
  # OTHERWISE RETURN -1
  def get(key)
    if @hash[key]
      node = @hash[key]
      @list.move_node_to_head(node)
      return node.data
    else
      return -1
    end
  end

  # ADD THE GIVEN KEY AND VALUE TO THE CACHE
  # IF THE CACHE ALREADY CONTAINS THE KEY, UPDATE ITS VALUE AND MOVE IT TO 
  # THE HEAD OF THE LIST
  # IF THE CACHE DOESN'T CONTAIN THE KEY, ADD IT TO THE CACHE AND PLACE IT
  # AT THE HEAD OF THE LIST
  # IF THE CACHE IS FULL, REMOVE THE LEAST RECENTLY USED ITEM BEFORE ADDING
  # THE NEW DATA TO THE CACHE
  def put(key, value)
    if @hash[key]
      node = @hash[key]
      node.data = value
      @list.move_node_to_head(node)
    else
      new_node = Node.new(value, key)
      @hash[key] = new_node
      @list.add_head(new_node)
      @size += 1

      if @size > @limit
        tail_node = @list.remove_tail
        @hash.delete(tail_node.key)
        @size -= 1
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  # Add your own tests here
end
Pseudocode for move_node_to_head method:
```
movenodeto_head(node):
if node is equal to the head:
return

remove_node(node)
add_head(node)

Explanation:
The `move_node_to_head` method in the `DoublyLinkedList` class moves a given node to the head of the list. It first checks if the given node is already the head. If it is, there is no need to move it, so the method returns early. Otherwise, it removes the node from its current position in the list by calling the `remove_node` method. Then, it adds the node to the head of the list by calling the `add_head` method, making the moved node the new head of the list.

In the `LRUCache` class, the `get` method retrieves the node from the cache using the key. If the node is in the cache, it moves the node to the head of the list using the `move_node_to_head` method and returns its data. If the node is not in the cache, it returns -1.

The `put` method adds the given key and value to the cache. If the cache already contains the key, it updates the value and moves the corresponding node to the head of the list. If the cache does not contain the key, it adds the key-value pair to the cache and places it at the head of the list. If the cache is full, it removes the least recently used item by calling the `remove_tail` method before adding the new data to the cache.

This implementation allows efficient access to the most recently used items by keeping them at the head of the list, while removing the least recently used items when the cache is full.

File: lru_cache.js
class Node {
  constructor(data = null, key = null, next = null, prev = null) {
    this.data = data;
    this.key = key;
    this.next = next;
    this.prev = prev;
  }
}

class DoublyLinkedList {
  constructor(head = null, tail = null) {
    this.head = head;
    this.tail = tail;
  }

  addHead(node) {
    // Add the given node to the head of the list
    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    }
  }

  removeTail() {
    // Remove the tail node from the list and return it
    if (!this.tail) {
      return null;
    } else {
      const tailNode = this.tail;
      if (this.head === this.tail) {
        this.head = null;
        this.tail = null;
      } else {
        this.tail.prev.next = null;
        this.tail = this.tail.prev;
      }
      return tailNode;
    }
  }

  removeNode(node) {
    // Remove the given node from the list and return it
    if (!node) {
      return;
    }

    if (node === this.head) {
      if (this.head === this.tail) {
        this.head = null;
        this.tail = null;
      } else {
        this.head = this.head.next;
        this.head.prev = null;
      }
    } else if (node === this.tail) {
      this.tail = this.tail.prev;
      this.tail.next = null;
    } else {
      node.prev.next = node.next;
      node.next.prev = node.prev;
    }

    node.next = null;
    node.prev = null;
  }

  moveNodeToHead(node) {
    // Move the given node from its location to the head of the list
    if (node === this.head) {
      return;
    }

    this.removeNode(node);
    this.addHead(node);
  }
}

class LRUCache {
  constructor(limit = 10) {
    this.limit = limit;
    this.size = 0;
    this.hash = {};
    this.list = new DoublyLinkedList(limit);
  }

  get(key) {
    // Retrieve the node from the cache using the key
    // If the node is in the cache, move it to the head of the list and return it
    // Otherwise, return -1
    if (this.hash[key]) {
      const node = this.hash[key];
      this.list.moveNodeToHead(node);
      return node.data;
    } else {
      return -1;
    }
  }

  put(key, value) {
    // Add the given key and value to the cache
    // If the cache already contains the key, update its value and move it to the head of the list
    // If the cache doesn't contain the key, add it to the cache and place it at the head of the list
    // If the cache is full, remove the least recently used item before adding the new data to the cache
    if (this.hash[key]) {
      const node = this.hash[key];
      node.data = value;
      this.list.moveNodeToHead(node);
    } else {
      const newNode = new Node(value, key);
      this.hash[key] = newNode;
      this.list.addHead(newNode);
      this.size++;

      if (this.size > this.limit) {
        const tailNode = this.list.removeTail();
        delete this.hash[tailNode.key];
        this.size--;
      }
    }
  }
}

if (require.main === module) {
  // Add your own tests here
}

module.exports = {
  Node,
  DoublyLinkedList,
  LRUCache
};
Pseudocode for moveNodeToHead method:
```
moveNodeToHead(node):
if node is equal to the head:
return

removeNode(node)
addHead(node)

//Explanation:
In the `LRUCache` class, the `get` method retrieves the node from the cache using the key. If the node is in the cache, it moves the node to the head of the list using the `moveNodeToHead` method and returns its data. If the node is not in the cache, it returns -1.

The `put` method adds the given key and value to the cache. If the cache already contains the key, it updates the value and moves the corresponding node to the head of the list. If the cache does not contain the key, it adds the key-value pair to the cache and places it at the head of the list. If the cache is full, it removes the least recently used item by calling the `removeTail` method before adding the new data to the cache.

This implementation allows efficient access to the most recently used items by keeping them at the head of the list, while removing the least recently used items when the cache is full.
